## Evaluation Projects

## Baseball Case Study

### Problem Statement:
This dataset utilizes data from 2014 Major League Baseball seasons in order to develop an algorithm that predicts the number of wins for a given team in the 2015 season based on several different indicators of success. There are 16 different features that will be used as the inputs to the machine learning and the output will be a value that represents the number of wins. <br>

-- Input features: Runs, At Bats, Hits, Doubles, Triples, Homeruns, Walks, Strikeouts, Stolen Bases, Runs Allowed, Earned Runs, Earned Run Average (ERA), Shutouts, Saves, Complete Games and Errors<br>

-- Output: Number of predicted wins (W)<br>

To understand the columns meaning, follow the link given below to understand the baseball statistics:<br> https://en.wikipedia.org/wiki/Baseball_statistics<br>

## Import Libraries

In [ ]:
# Analysis libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import zscore,boxcox

# Machine learning libraries
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import power_transform, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import warnings
warnings.filterwarnings('ignore')

In [ ]:
df = pd.read_csv('baseball.csv')
df.head()

<b>Linear Regression</b> Model needs to be used for this dataset

In [ ]:
df.columns

## Description of Dataset

Runs, At Bats, Hits, Doubles, Triples, Homeruns, Walks, Strikeouts, Stolen Bases, Runs Allowed, Earned Runs, Earned Run Average (ERA), Shutouts, Saves, and Errors <br>

   <b>Input features:</b><br>
    R  - Runs<br>
    AB - At Bats<br>
    H  - Hits<br>
    2B - Doubles<br>
    3B - Triples<br>
    HR - Homeruns<br>
    BB - Walks<br>
    SO - Strikeouts<br>
    SB - Stolen Bases<br>
    RA - Run Allowed<br>
    ER - Earned Runs<br>
    ERA - Earned Run Average (ERA)<br>
    CG - Complete games ( referred online)<br>
    SHO - Shutouts<br>
    SV - Saves<br>
    E - Errors<br>
  ----------------------------------------------------<br>  
 <b>Output features:</b><br>
    W -Number of predicted wins (W)<br>

In [ ]:
# size of dataset
df.shape

 - There are 30 rows and 17 columns

In [ ]:
df.dtypes

 - All values are in numeric only so need to use label encoder.
 - There are 16 integers and 1 float present in dataset.
 - There are no categorical values in dataset.

In [ ]:
# Check null values
df.isnull().sum()

 - No null values in dataset.

In [ ]:
# recheck null
sns.heatmap(df.isnull())

 - Recheck null, showing no null values in dataset

## EDA

In [ ]:
## Statistical summary
df.describe()

 - Some of the columns having outliers and skewness, have to observe that.

## Observation :

In [ ]:
df.iloc[:,0:10].plot(kind='box',subplots=True,layout=(2,5))

In [ ]:
df.iloc[:,10:].plot(kind='box',subplots=True,layout=(2,5))

 - R,ERA,SHO,SV have outliers.
 - Other colums don't have

## Univarient Analysis

In [ ]:
df['E'].plot(kind='box')

In [ ]:
df['R'].plot(kind='box')

In [ ]:
# checking skewness
df.skew()

 - As per standards(<0.5), there are skewness presnt in R,H,CG,SHO,SV,E
 - We need to handle the skewness.

In [ ]:
# checking distribution plot for more detail

for i in df.columns:
    plt.figure()
    sns.distplot(df[i])

## Checking correlation of data

 - As per the above plot R,H,HR,SO,SHO,CG,E having some skewness

In [ ]:
plt.figure(figsize=(12,10))
sns.heatmap(df.corr(),annot=True)

 - RA,ER,ERA are highly -vely correlated with target variable

## Bavarient Analysis

In [ ]:
plt.scatter(df.R,df.W)
plt.xlabel('Runs')
plt.ylabel('Wins')
plt.show()

In [ ]:
# runs relation with dataset
for i in df.columns:
    plt.figure()
    sns.barplot(x=df[i],y=df.W)

 - Relationship between features and label have been observed.

In [ ]:
## Removing outliers from data
z = np.abs(zscore(df))
z

In [ ]:
df_new = df[(z<3).all(axis=1)]

In [ ]:
df.shape

In [ ]:
# checking shape of new dataset
df_new.shape

## Removing highly negative correlated columns

In [ ]:
df_new.drop(['RA','ER'],axis=1,inplace=True)

In [ ]:
df_new.drop(['ERA',],axis=1,inplace=True)

## Splitting X & Y data

In [55]:
x = df_new.iloc[:,1:-1]
y = df_new['W']

In [56]:
x

R    AB     H   2B  3B   HR   BB    SO   SB  CG  SHO  SV
0   724  5575  1497  300  42  139  383   973  104   2    8  56
1   696  5467  1349  277  44  156  439  1264   70   2   12  45
2   669  5439  1395  303  29  141  533  1157   86  11   10  38
3   622  5533  1381  260  27  136  404  1231   68   7    9  37
4   689  5605  1515  289  49  151  455  1259   83   7   12  35
6   764  5567  1397  272  19  212  554  1227   63   3    4  48
7   713  5485  1370  246  20  217  418  1331   44   0   10  43
8   644  5485  1383  278  32  167  436  1310   87   1   12  60
9   748  5640  1495  294  33  161  478  1148   71   3   10  40
10  751  5511  1419  279  32  172  503  1233  101   5    9  45
11  729  5459  1363  278  26  230  486  1392  121   5   13  39
12  661  5417  1331  243  21  176  435  1150   52   2   12  46
13  656  5544  1379  262  22  198  478  1336   69   6   12  45
14  694  5600  1405  277  46  146  475  1119   78   5   15  28
15  647  5484  1386  288  39  137  506  1267   69   1   15  62
16  697  5631  1462  292  27  140  461  1322   98   0   13  54
17  689  5491  1341  272  30  171  567  1518   95   6   21  48
18  655  5480  1378  274  34  145  412  1299   84   1    7  40
19  640  5571  1382  257  27  167  496  1255  134   2    8  35
20  683  5527  1351  295  17  177  488  1290   51   1   14  50
21  703  5428  1363  265  13  177  539  1344   57   4   13  41
22  613  5463  1420  236  40  120  375  1150  112   0   12  35
23  573  5420  1361  251  18  100  471  1107   69   3   10  44
24  626  5529  1374  272  37  130  387  1274   88   1    7  35
25  667  5385  1346  263  26  187  563  1258   59   6   21  47
26  696  5565  1486  288  39  136  457  1159   93   7   18  41
27  720  5649  1494  289  48  154  490  1312  132   1   12  44
28  650  5457  1324  260  36  148  426  1327   82   1    6  41
29  737  5572  1479  274  49  186  388  1283   97   4    4  36

In [57]:
y

0      95
1      83
2      81
3      76
4      74
6      87
7      81
8      80
9      78
10     88
11     86
12     85
13     76
14     68
15    100
16     98
17     97
18     68
19     64
20     90
21     83
22     71
23     67
24     63
25     92
26     84
27     79
28     74
29     68
Name: W, dtype: int64

## Removing Skewness

In [58]:
df_x = power_transform(x)
df_x = pd.DataFrame(df_x)

In [59]:
x = df_x
x

0    1    2         3         4         5         6         7   \
0   0.962544  0.0  0.0  1.685188  1.006150 -0.741927 -1.605198 -2.550612   
1   0.298863  0.0  0.0  0.138198  1.185227 -0.109958 -0.462096  0.093683   
2  -0.312105  0.0  0.0  1.907385 -0.228819 -0.664354  1.232098 -0.935611   
3  -1.308298  0.0  0.0 -0.837665 -0.432228 -0.860039 -1.162721 -0.230683   
4   0.137737  0.0  0.0  0.911435  1.622636 -0.289647 -0.155686  0.044143   
5   1.964209  0.0  0.0 -0.163010 -1.295827  1.631637  1.579494 -0.269583   
6   0.698146  0.0  0.0 -1.542635 -1.182758  1.767734 -0.877217  0.770980   
7  -0.852595  0.0  0.0  0.199897  0.068703  0.269125 -0.520476  0.556008   
8   1.555951  0.0  0.0  1.255256  0.166017  0.065014  0.270944 -1.019210   
9   1.631727  0.0  0.0  0.262086  0.068703  0.434620  0.717576 -0.211199   
10  1.084297  0.0  0.0  0.199897 -0.535590  2.109761  0.415731  1.409203   
11 -0.487692  0.0  0.0 -1.682709 -1.071323  0.564164 -0.540002 -1.000673   
12 -0.596176  0.0  0.0 -0.729847 -0.961425  1.235848  0.270944  0.822524   
13  0.252632  0.0  0.0  0.138198  1.361887 -0.474340  0.216181 -1.285401   
14 -0.789011  0.0  0.0  0.844222  0.732678 -0.820431  0.770046  0.123475   
15  0.322037  0.0  0.0  1.116168 -0.432228 -0.703026 -0.042862  0.678550   
16  0.137737  0.0  0.0 -0.163010 -0.128676  0.401843  1.789632  2.791487   
17 -0.617757  0.0  0.0 -0.043976  0.262460 -0.511906 -0.998685  0.444379   
18 -0.936834  0.0  0.0 -0.996003 -0.432228  0.269125  0.594227  0.004612   
19  0.001150  0.0  0.0  1.325587 -1.527310  0.596168  0.451649  0.353547   
20  0.461904  0.0  0.0 -0.564688 -2.015649  0.596168  1.332385  0.905280   
21 -1.489372  0.0  0.0 -1.995042  0.824510 -1.529112 -1.778543 -1.000673   
22 -2.256817  0.0  0.0 -1.300683 -1.410638 -2.478298  0.142759 -1.394124   
23 -1.226825  0.0  0.0 -0.163010  0.546878 -1.102893 -1.519540  0.193184   
24 -0.356235  0.0  0.0 -0.675253 -0.535590  0.908214  1.725362  0.034252   
25  0.298863  0.0  0.0  0.844222  0.732678 -0.860039 -0.117955 -0.916971   
26  0.865849  0.0  0.0  0.911435  1.536265 -0.181252  0.487457  0.576376   
27 -0.725080  0.0  0.0 -0.837665  0.452861 -0.399842 -0.717273  0.729845   
28  1.281152  0.0  0.0 -0.043976  1.622636  0.877642 -1.498230  0.283212   

          8         9         10        11  
0   0.936132 -0.307098 -0.787002  1.532753  
1  -0.516377 -0.307098  0.236737  0.312020  
2   0.225038  2.011315 -0.252844 -0.664137  
3  -0.618422  1.264463 -0.513555 -0.820689  
4   0.095038  1.264463  0.236737 -1.149165  
5  -0.884526  0.121871 -2.064039  0.677176  
6  -2.082843 -1.732896 -0.252844  0.052325  
7   0.267558 -0.870682  0.236737  1.908137  
8  -0.466233  0.121871 -0.252844 -0.365006  
9   0.824915  0.770649 -0.513555  0.312020  
10  1.521413  0.770649  0.468029 -0.512328  
11 -1.535760 -0.307098  0.236737  0.436838  
12 -0.567102  1.031663  0.236737  0.312020  
13 -0.130312  0.770649  0.908147 -2.498840  
14 -0.567102 -0.870682  0.908147  2.084917  
15  0.710994 -1.732896  0.468029  1.333006  
16  0.594218  1.031663  2.090856  0.677176  
17  0.138787 -0.870682 -1.075340 -0.365006  
18  1.925528 -0.307098 -0.787002 -1.149165  
19 -1.600293 -0.870682  0.691583  0.906024  
20 -1.227262  0.472300  0.468029 -0.221934  
21  1.220510 -1.732896  0.236737 -1.149165  
22 -0.567102  0.121871 -0.252844  0.183904  
23  0.309686 -0.870682 -1.075340 -1.149165  
24 -1.109925  1.031663  2.090856  0.558509  
25  0.514700  1.264463  1.521828 -0.221934  
26  1.865486 -0.870682  0.236737  0.183904  
27  0.050861 -0.870682 -1.381412 -0.221934  
28  0.672394  0.472300 -2.064039 -0.982264

## Standard Scaling

In [60]:
scaler = StandardScaler()
x = scaler.fit_transform(x)

In [61]:
x

array([[ 9.62543504e-01,  0.00000000e+00,  0.00000000e+00,
         1.68518793e+00,  1.00615029e+00, -7.41927000e-01,
        -1.60519802e+00, -2.55061247e+00,  9.36131648e-01,
        -3.07098204e-01, -7.87002186e-01,  1.53275292e+00],
       [ 2.98863300e-01,  0.00000000e+00,  0.00000000e+00,
         1.38197902e-01,  1.18522654e+00, -1.09958425e-01,
        -4.62095966e-01,  9.36832915e-02, -5.16377335e-01,
        -3.07098204e-01,  2.36736538e-01,  3.12020186e-01],
       [-3.12105130e-01,  0.00000000e+00,  0.00000000e+00,
         1.90738550e+00, -2.28819392e-01, -6.64354121e-01,
         1.23209786e+00, -9.35611465e-01,  2.25038365e-01,
         2.01131531e+00, -2.52844176e-01, -6.64136739e-01],
       [-1.30829774e+00,  0.00000000e+00,  0.00000000e+00,
        -8.37664770e-01, -4.32227907e-01, -8.60039342e-01,
        -1.16272085e+00, -2.30682707e-01, -6.18421529e-01,
         1.26446344e+00, -5.13554932e-01, -8.20688859e-01],
       [ 1.37737301e-01,  0.00000000e+00,  0.0000000

## Finding Best random state

In [66]:
maxScore = 0
macRS = 0
for i in range(1,200):
    x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.23,random_state=i)
    lr=LinearRegression()
    lr.fit(x_train,y_train)
    pred_train = lr.predict(x_train)
    pred_test = lr.predict(x_test)
    print(f'Best Random State : {i} and Train R2 Score : {r2_score(y_train,pred_train)}')
    print(f'Best Random State : {i} and Test R2 Score : {r2_score(y_test,pred_test)}')
    print('')

Best Random State : 1 and Train R2 Score : 0.8550783206268826
Best Random State : 1 and Test R2 Score : 0.6944530201712394

Best Random State : 2 and Train R2 Score : 0.8173961865951549
Best Random State : 2 and Test R2 Score : -0.8302730889884651

Best Random State : 3 and Train R2 Score : 0.8199055316941919
Best Random State : 3 and Test R2 Score : 0.7257992916491386

Best Random State : 4 and Train R2 Score : 0.8536067232046677
Best Random State : 4 and Test R2 Score : 0.618508188589969

Best Random State : 5 and Train R2 Score : 0.8806333789341638
Best Random State : 5 and Test R2 Score : 0.2671195820874026

Best Random State : 6 and Train R2 Score : 0.8800922107276878
Best Random State : 6 and Test R2 Score : 0.6629688959955353

Best Random State : 7 and Train R2 Score : 0.8488482103041881
Best Random State : 7 and Test R2 Score : 0.38500783612986544

Best Random State : 8 and Train R2 Score : 0.8368074421041336
Best Random State : 8 and Test R2 Score : 0.7888158307008533

Best Ra

<b>Best Random State : 23 and Train R2 Score : 0.8273528079101595<br>
        <br>
Best Random State : 23 and Test R2 Score : 0.7812319076407748</b>

## Splitting test and training data

In [68]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.23,random_state=23)

## LinearRegression

In [69]:
lr = LinearRegression()
lr.fit(x_train,y_train)
pred = lr.predict(x_test)

In [75]:
## Performance of model
def performance(actual,pred):
    print('Error ==> ')
    print('Mean Absolute error : ',mean_absolute_error(actual,pred))
    print('Mean Squared error : ',mean_squared_error(actual,pred))
    print('R2_Score : ',r2_score(actual,pred))

In [76]:
performance(y_test,pred)

Error ==> 
Mean Absolute error :  4.388362740359034
Mean Squared error :  30.797189818243584
R2_Score :  0.7812319076407748


## Cross Validation

In [77]:
from sklearn.model_selection import cross_val_score

In [79]:
cross = cross_val_score(LinearRegression(),x,y,cv=10)
cross.mean()

-5.944004960197801

## Regularization

In [81]:
from sklearn.linear_model import Lasso,Ridge,ElasticNet

In [83]:
# We can corrected by alpha value
le = Lasso(alpha=0.0001)
le.fit(x_train,y_train)
predict = le.predict(x_test)
le.score(x_train,y_train)

print(r2_score(y_test,predict))

0.7812746478132153


 - The score is 78%

## Cross Validation

In [84]:
cross_val = cross_val_score(le,x,y,cv=10)
cross_val.mean()

-5.943691740647567

In [85]:
# Ridge Model
re = Ridge(alpha=0.0001)
re.fit(x_train,y_train)
re.score(x_train,y_train)

0.8273528077486438

 - Ridge is giving 82%

## Cross Validation

In [86]:
cross_val = cross_val_score(re,x,y,cv=10)
cross_val.mean()

-5.943965765736464

## Model

In [87]:
from sklearn.svm import SVR

kernel = ['linear','poly','rbf','sigmoid']

for i in kernel:
    sr = SVR(kernel=i)
    sr.fit(x_train,y_train)
    print(sr.score(x_train,y_train))
    prec = sr.predict(x_test)
    print('Erro ==>')
    print('Mean Absolute error : ',mean_absolute_error(y_test,prec))
    print('Mean Squared error : ',mean_squared_error(y_test,prec))
    cross_val = cross_val_score(sr,x,y,cv=10)
    print('Cross Validation : ',cross_val.mean())
    print('')

0.7179715625258282
Erro ==>
Mean Absolute error :  4.987851266476606
Mean Squared error :  34.70558461989886
Cross Validation :  -7.021812327196119

0.3586794381170487
Erro ==>
Mean Absolute error :  9.496217537067833
Mean Squared error :  134.59662762091347
Cross Validation :  -5.847696092613174

0.17746992568223285
Erro ==>
Mean Absolute error :  9.199110897190863
Mean Squared error :  130.74143458890723
Cross Validation :  -6.54114430244923

0.27731531654048813
Erro ==>
Mean Absolute error :  7.788038340251693
Mean Squared error :  102.80343305349395
Cross Validation :  -4.160333456071724



 - This is giving 71% and less error for linear model.

## Hyper Parameter Tuning 

In [89]:
from sklearn.model_selection import GridSearchCV

In [124]:
# lasso Regression param

param = {'alpha' :[0.1,0.01,0.001,0.0001],
         'selection' :['cyclic','random']}

In [138]:
lasso_hp = GridSearchCV(Lasso(),param,cv=5)
lasso_hp.fit(x_train,y_train)
lasso_hp.best_params_

{'alpha': 0.1, 'selection': 'random'}

In [139]:
# By using alpha value

le = Lasso(alpha=0.1,selection='random')
le.fit(x_train,y_train)
predict = le.predict(x_test)
le.score(x_train,y_train)

print(r2_score(y_test,predict))

0.8041130755251142


In [127]:
# Ridge Regression Param

param_r = {'alpha' :[0.1,0.01,0.001,0.0001],
           'solver' :['auto','svd','cholesky','lsqr','sparse_cg','sag','saga']}

In [128]:
ridge_hp = GridSearchCV(Ridge(),param_r,cv=5)
ridge_hp.fit(x_train,y_train)
ridge_hp.best_params_

{'alpha': 0.1, 'solver': 'saga'}

In [129]:
# Ridge Model

re = Ridge(alpha=0.1,solver='saga')
re.fit(x_train,y_train)
predict = le.predict(x_test)
re.score(x_train,y_train)

print(r2_score(y_test,predict))

0.8041590369937879


 - Both Lasso & Ridge giving 80.4% accuracy.

## Save the Model

In [132]:
import joblib

joblib.dump(le,'Base_ball_case_study.obj')

['Base_ball_case_study.obj']

In [133]:
joblib.dump(re,'Base_ball_case_study1.obj')

['Base_ball_case_study1.obj']

## Load the model to file

In [140]:
mod1 = joblib.load('Base_ball_case_study.obj')
mod1

Lasso(alpha=0.1)

In [136]:
mod1 = joblib.load('Base_ball_case_study1.obj')
mod1

Ridge(alpha=0.1, solver='saga')